lib/auth/auth.go

```diff
@@ -2541,6 +2562,183 @@ func (a *Server) DeleteNetworkRestrictions(ctx context.Context) error {
 	return a.Services.Restrictions.DeleteNetworkRestrictions(ctx)
 }
 
+// newKeySetForKeySet returns a new CAKeySet with local keys of the same types
+// (SSH, TLS, or JWT) as the argument keySet
+func newKeySetForKeySet(keyStore keystore.KeyStore, clusterName string, keySet types.CAKeySet) (types.CAKeySet, error) {
+	var newKeySet types.CAKeySet
+	if len(keySet.SSH) != 0 {
+		sshKeyPair, err := keyStore.NewSSHKeyPair()
+		if err != nil {
+			return newKeySet, trace.Wrap(err)
+		}
+		newKeySet.SSH = append(newKeySet.SSH, sshKeyPair)
+	}
+	if len(keySet.TLS) != 0 {
+		tlsKeyPair, err := keyStore.NewTLSKeyPair(clusterName)
+		if err != nil {
+			return newKeySet, trace.Wrap(err)
+		}
+		newKeySet.TLS = append(newKeySet.TLS, tlsKeyPair)
+	}
+	if len(keySet.JWT) != 0 {
+		jwtKeyPair, err := keyStore.NewJWTKeyPair()
+		if err != nil {
+			return newKeySet, trace.Wrap(err)
+		}
+		newKeySet.JWT = append(newKeySet.JWT, jwtKeyPair)
+	}
+	return newKeySet, nil
+}
+
+func mergeKeySets(a, b types.CAKeySet) types.CAKeySet {
+	newKeySet := a.Clone()
+	newKeySet.SSH = append(newKeySet.SSH, b.SSH...)
+	newKeySet.TLS = append(newKeySet.TLS, b.TLS...)
+	newKeySet.JWT = append(newKeySet.JWT, b.JWT...)
+	return newKeySet
+}
+
+// atomicCAUpdate performs an atomic CompareAndSwap to update the given CA with
+// newKeys. getKeySet and setKeySet should be passed to choose between the
+// active keyset and the additional trusted keyset.
+func (a *Server) atomicCAUpdate(
+	currentCA types.CertAuthority,
+	newKeys types.CAKeySet,
+	getKeySet func(types.CertAuthority) types.CAKeySet,
+	setKeySet func(types.CertAuthority, types.CAKeySet) error,
+	needsUpdate func(types.CertAuthority) bool) error {
+	for {
+		select {
```

Comment:
It might, do you have a suggestion for a reasonably small (and I'm assuming randomized) amount of time? I was thinking collisions would be pretty rare here and the number of auth servers competing will be pretty small (the rotation poll period is already jittered)